proposals/text/5540-llm-support-2.0.md

```diff
@@ -0,0 +1,309 @@
+
+Due to the dynamic nature of prompt templates, the `Prompts.run()`  method takes `kwargs`, which contains all the variables that will be filled in the template. However, for this component to work with Canals, we need to know in advance which values this dict will contain: therefore, we need the users to specify in the `__init__` of the component which parameters to expect in the template.
+
+Keep in mind that such parameters names **cannot be changed at runtime**.
+
+We foresee `Prompts` to output a list of prompts, not necessarily just one, as currently  some of our existing prompt templates produce several output prompts when rendered.
+
+Draft I/O for `Prompts`:
+
+```python
+@component
+class Prompts:
+
+    def __init__(self, template_variables: Union[str, Path]):
+        self.template_variables = template_variables
+		component.set_input_parameters(**{var: Any for var in template_variables})
+
+  	@component.output_types(prompts=List[str])
+    def run(self, template: str, **kwargs):
+        self.template_text = # Load the template
+
+        variables = # extracts the variables from the template text
+        if variables != self.template_variables:
+            raise ValueError()
+
+        # Render the template using the variables
+        return {"prompts": prompts}
+```
+
+### **Why we need to specify the variables in `__init__`?**
+
+The design above derives from one Canals limitation: component’s sockets need to be all known the latest at `__init__` time, in order for the connections to be made and validated. Therefore, we need to know all the prompt variables before building the pipelines, because the prompt variables are inputs of the `run()` method.
+
+However, earlier iterations of Canals did support so-called “true variadic” components: components that do not need to know what they will be connected to, and build the input sockets at need. Such components of course lack input validation, but enable usecases like the above.
+
+If we decide that Canals should support again such components, we would be able to rewrite `Prompts` to take a prompt as its input parameter and just accept any other incoming input, on the assumption that users knows that they’re doing.
+
+For example:
+
+```python
+@component
+class Prompts:
+
+	@variadic_input
+  	@component.output_types(prompts=List[str])
+    def run(self, template: Union[str, Path], **kwargs):
+	    # ... loads the template ...
+        # ... render the prompts ...
+        return {"prompts": prompts}
+```
+
+### Why a separate `Prompts` component at all?
+
+`PromptNode` used to take the prompt template and the variables to render it directly, and then forward the result to the LLM.
+
+**PRO: Allows using various templating and guidance libraries, or none at all**
+
+The key advantage of `Prompts` is ability to use any tool from the ever growing list of LLM prompting template libs. These include but are not limited to: https://github.com/microsoft/guidance,
```

Comment:
I don't see how this would support guidance? Guidance allows to code the control flow for LLM generation within the template. Instructions can be followed by macros that actually call the LLM, then generate something, then the next part of the prompt will be added and the LLM might be called again.

Reply:
This topic will definitely see more discussion when we design Agents as well. I put it into here to clarify why I consider separate Prompt components valuable, but if you find it confusing or really wrong I'll take this paragraph out and move the discussion to the Agents proposal.

Reply:
`guidance` is not the example as Mathis says. 

The idea is that since prompt components will take the already formatted prompt as input, and the rendering is delegated to another component, users can create their renderer component. Maybe they want to use a custom proprietary template language, or just any other template language.

This new architecture enables this use case.